File: 02 백준/08 정렬/Python/Prob20920.py

"""
title  : 20920. 영단어 암기는 괴로워 (Silver 3)
time   : 300ms        472ms
memory : 59740KB      140824KB
"""

import sys

def solution():
    input = sys.stdin.readline

    N, M = map(int, input().split())

    word = dict()

    for _ in range(N):
        w = input().rstrip()

        if M <= len(w):
            word[w] = word.get(w, 0) + 1

    for (k, _) in sorted(word.items(), key=lambda x: (-x[1], -len(x[0]), x[0])):
        print(k)

solution()

File: 02 백준/08 정렬/Java/src/Prob20920.java

/**
* title  : 20920. 영단어 암기는 괴로워 (Silver 3)
* time   : 684ms
* memory : 37712KB
*/

import java.io.*;
import java.util.*;

public class Prob20920 {
    
    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        HashMap<String, Integer> wordMap = new HashMap<>();
        for (int i = 0; i < N; i++) {
            String word = br.readLine();

            if (M <= word.length()) {
                wordMap.put(word, wordMap.getOrDefault(word, 0) + 1);
            }
        }

        List<String> keySet = new ArrayList<>(wordMap.keySet());
        
        keySet.sort(new Comparator<String>() {
            @Override
            public int compare(String o1, String o2) {
                if (wordMap.get(o1).equals(wordMap.get(o2))) {
                    if (o1.length() == o2.length()) {
                        return o1.compareTo(o2);
                    }
                    return o2.length() - o1.length();
                }
                return wordMap.get(o2).compareTo(wordMap.get(o1));
            }   
        });

        for (String key : keySet) {
            sb.append(key).append("\n");
        }

        System.out.println(sb);
    }
}
